### DISCUSSION TITLE
Thread Pool Size

### DISCUSSION THREAD

**Discussion by lf-jason-jia** [Kong/unirest-java#459] [Q&A] [answered]
Do we have a way of changing the size the underlying thread pool?

When I run this code, I got "2 2":
```java
    @Test
    void basic_unirest_sync() {
        System.out.println(Thread.activeCount());
        Object res = Unirest.get("https://reqres.in/api/users?page=2");
        System.out.println(Thread.activeCount());
        assertNotNull(res);
    }
```

But if I run the async version, I got "2 16":
```java
    @Test
    void basic_unirest_async() {
        System.out.println(Thread.activeCount());
        Object res = Unirest.get("https://reqres.in/api/users?page=2").asObjectAsync(Object.class).join();
        System.out.println(Thread.activeCount());
        assertNotNull(res);
    }
```

Please correct me if my understanding is incorrect but I think the whole point of using async is to minimize the use of threads while providing a non-blocking API? I assume an async task could be suspended and resumed so in theory we only need a couple of threads to offer an async API?

**Comment by ryber** [comment, marked as answer]
What version of Unirest are you using?

**Reply by samadpaydar3** [reply]
I have the same issue with version 3.13.10

**Reply by ryber** [reply]
3 and 4 are very different. 

3 Uses ApacheHttpClient for it's engine, it also starts a special thread that sweeps the client for hung connections and kills them. This thread is not involved in actual requests and is not very active. As for what Apache does inside, you would have to ask them, but my understanding is there is an event loop in there dequeuing io operations, but it may use more than 1.

4 Uses the native Java 11 HttpClient. This does not have the watcher thread, the Java Client is fully reactive, but it does have a thread executor, which you can override in the Unrest config 

see https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.Builder.html#executor(java.util.concurrent.Executor)

The Unirest exposure of this is like:
```
Unirest.config().executor(Executors.newFixedThreadPool(1))
```

**Reply by samadpaydar3** [reply]
Thanks for the info. I will try to upgrade to 4 to see how it affects my code.
